pkg/controllers/deprovisioning/types.go

```diff
@@ -0,0 +1,160 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deprovisioning
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/metrics"
+
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+)
+
+// Result is used to indicate the action of consolidating so we can optimize by not trying to consolidate if
+// we were unable to consolidate the cluster and it hasn't changed state with respect to pods/nodes.
+type Result byte
+
+const (
+	ResultNothingToDo Result = iota // there are no actions that can be performed given the current cluster state
+	ResultRetry                     // we attempted an action, but its validation failed so retry soon
+	ResultFailed                    // the action failed entirely
+	ResultSuccess                   // the action was successful
+)
+
+func (r Result) String() string {
+	switch r {
+	case ResultNothingToDo:
+		return "Nothing to do"
+	case ResultRetry:
+		return "Retry"
+	case ResultFailed:
+		return "Failed"
+	case ResultSuccess:
+		return "Success"
+	default:
+		return fmt.Sprintf("Unknown (%d)", r)
+	}
+}
+
+type deprovisioner interface {
```

Comment:
My sense is that this is a bit rigid. I haven't tried factoring it myself, but it's rare that any abstraction has so many fundamental characteristics. Often, when interfaces have this many specialized concepts and parameters, they don't survive very long as requirements change. Happy to start here and iterate, though.

Reply:
I can understand your thoughts because there's a lot of functions. I wanted to follow the same structure that Consolidation uses, since it takes a lot pre-caution and has all the general deprovisioning requirements. I pulled out all `deprovisioner` specific logic here into an interface, and I agree it'll definitely improve and change in the future, but it follows the design we discussed.

Reply:
Do you have a sense of what the next iteration of this interface might look like? No idea if this is even close, but squinting at it:
```
type deprovisioner interface {
  // returns sorted
  GetCandidateActions() []action
}
type action interface {
  Target() []*v1.Node
  Destination() []*v1.Node
  // Check that the action is still valid against an updated cluster state
  Validate(*state.Cluster) error
}
```

Reply:
Not sure, but the current implementation is pretty particular to save computation, but I could see a case for simplifying it if none of the other controllers require any fancy logic.

Reply:
I don't think we should sacrifice performance for simplicity, but it's worth thinking about how to achieve both.